pkg/lib/csv/notification.go

```diff
@@ -0,0 +1,17 @@
+package csv
+
+import (
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+)
+
+// WatchNotification is an sink interface that can be used to get notification
+// of CSV reconciliation request(s) received by the operator.
+type WatchNotification interface {
```

Comment:
This seems to overlap with [`cache.ResourceEventHandler`](https://godoc.org/k8s.io/client-go/tools/cache#ResourceEventHandler), multiple of which can be [registered to an informer](https://github.com/kubernetes/client-go/blob/7b18d6600f6b0022e31c46b46875beffd85cc71a/tools/cache/shared_informer.go#L326). Adding a "CVO notifying" `cache.ResourceEventHandler` to our [CSV informers in `NewOperator(...)`](https://github.com/operator-framework/operator-lifecycle-manager/blob/0a281bc71f56fc25cab8f6e318499160652e97b6/pkg/controller/operators/olm/operator.go#L264) might look something like:

```go
// ...
csvInformer := sharedInformerFactory.Operators().V1alpha1().ClusterServiceVersions()
// Add handler for updating ClusterStatus on packageserver CSV changes
csvInformer.Informer().AddEventHandler(op.packageServerClusterStatusHandler)
// ...
```

This gives us a single point of entry for notifying CVO, rather than adding update calls to sync handlers individually. I may not be getting the full picture here though, do you think using `cache.ResourceEventHandler` makes sense?

Reply:
This is the requirement:
We need to be notified of all CSV resource(s) being watched by olm operator. What we need is an interface that injects an `*v1alpha1.ClusterServiceVersion` object whenever it is being reconciled. 

Given this, I thought tapping into `syncClusterServiceVersion` would allow us to send these notification(s) without knowing how the informers/indexers/cache are being set up. This way we can keep the notification logic decoupled from the informers. We also can afford a statically `typed` interface here, the object we receive is of `*v1alpha1.ClusterServiceVersion` type. This also makes it limited to `ClusterServiceVersion` type only as you mentioned.

I am still learning about informer stuffs, but I thought `cache.ResourceEventHandler` funcs sometimes may provide you with the key of the object rather than the object itself.

We can use `cache.ResourceEventHandler` interface to make it applicable to all types. Thinking out loud here, let's say the `Operator` exposes a function `RegisterWatchedResourceNotification(obj interface{}, handlers ResourceEventHandler)`.
`obj` is they type of resources we want to get notification of, it could be `Subscription` or `ClusterServiceVersion`. 

I am happy to make it more generic, thinking whether to do it as part of this PR or a separate one.
Thoughts @njhale?

Reply:
I was kind of thinking the same thing. I don't know that a statically typed interface outweighs having all the event watcher logic in one place. But I'm fairly certain that a follow up PR for such a change is fine.

Reply:
@tkashem I see where you're coming from and I definitely think using a statically typed interface is a good thing for high-level consumers - this also lines up with [some of our thoughts for status updates](https://docs.google.com/document/d/1ROc7kDkZE57tS_Ttx5xf3_aIQLVbMsckg3YSQnVvFvU/edit?usp=sharing).

> I thought cache.ResourceEventHandler funcs sometimes may provide you with the key of the object rather than the object itself.

Correct, on add and update a key is given that does not include the new object. On Delete events, the `cache.ResourceEventHandler` receives a "tombstone" that contains the object - Our current `QueueInformer` abstractions handle [enqueuing keys given to a default event handler onto a workqueue](https://github.com/operator-framework/operator-lifecycle-manager/blob/41821c7a460ea13ee421a57f4d8cc4bb1a356a64/pkg/lib/queueinformer/queueinformer.go#L62) and [consuming keys from that queue](https://github.com/operator-framework/operator-lifecycle-manager/blob/41821c7a460ea13ee421a57f4d8cc4bb1a356a64/pkg/lib/queueinformer/queueinformer_operator.go#L175). They also use those keys to look up resources in the backing cache and [feeds them to a registered `syncHandler` function](https://github.com/operator-framework/operator-lifecycle-manager/blob/41821c7a460ea13ee421a57f4d8cc4bb1a356a64/pkg/lib/queueinformer/queueinformer_operator.go#L205).

I definitely think there's room for improvement in what we currently have. I just want to be careful of introducing more parallel abstractions. I would like to consolidate your ideas, `QueueInformer`, and our other thoughts, and come up with something small, sharp, and simple that we can apply across the board. This can definitely be an incremental thing as you and @jpeeler have suggested.

Reply:
@njhale We are are in agreement. I did see room for improvement on how we currently build our operator instances. I decided not to partake into any refactoring, I just did the minimum for my PR to work :). 

I skimmed through your refactor on the operator builder, I think we are heading in the right direction.